test/e2e/util_gatewayapi_test.go

```diff
@@ -590,6 +643,132 @@ func assertGatewaySuccessful(t *testing.T, namespace, name string) (*gatewayapiv
+// The expectations parameter is a map where keys are expectations for DNSRecord and values indicate whether a DNSRecord should be present.
+func assertExpectedDNSRecords(t *testing.T, expectations map[expectedDnsRecord]bool) error {
+	t.Helper()
+
+	var expectationsMet bool
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		haveExpectNotPresent := false
+		// expectationsMet starts true and gets set to false when some expectation is not met.
+		expectationsMet = true
+
+		dnsRecords := &v1.DNSRecordList{}
+		if err := kclient.List(context, dnsRecords, client.InNamespace(operatorcontroller.DefaultOperandNamespace)); err != nil {
+			return false, fmt.Errorf("failed to list DNSRecords: %v", err)
+		}
+
+		// Iterate over all expectations.
+		for exp, shouldBePresent := range expectations {
+			if !shouldBePresent {
+				haveExpectNotPresent = true
+			}
+
+			// Reset the found and foundReady flags for each expectation.
+			found := false
+			foundReady := false
+			// Look for a DNSRecord that matches the expected gateway and DNS name.
+			for _, record := range dnsRecords.Items {
+				if record.Labels["gateway.networking.k8s.io/gateway-name"] == exp.gatewayName &&
+					record.Spec.DNSName == exp.dnsName {
+
+					if !shouldBePresent {
+						expectationsMet = false
+						found = true
+						t.Logf("DNSRecord %q (%s) found but should not be present.", record.Name, exp.dnsName)
+						return false, nil
+					}
+					found = true
+					// DNSRecord found and should be present, check if it is published
+					for _, zone := range record.Status.Zones {
+						for _, condition := range zone.Conditions {
+							if condition.Type == v1.DNSRecordPublishedConditionType && condition.Status == string(metav1.ConditionTrue) {
+								t.Logf("Found DNSRecord %q (%s) %s=%s as expected", record.Name, exp.dnsName, condition.Type, condition.Status)
+								foundReady = true
+							}
+						}
+					}
+					if !foundReady {
+						t.Logf("Found DNSRecord %v but could not determine its readiness; retrying...", record.Name)
+						expectationsMet = false
+						return false, nil
+					}
+				}
+			}
+
+			// If the record is expected but not found, return false to continue polling.
+			if shouldBePresent && !found {
+				t.Logf("DNSRecord for hostname %q (gateway: %s) is expected to be present but was not found; retrying...", exp.dnsName, exp.gatewayName)
+				expectationsMet = false
+				return false, nil
+			}
+			// If the record is not expected but was found, return false to continue polling.
+			if !shouldBePresent && found {
+				t.Logf("DNSRecord for hostname %q (gateway: %s) is present but was expected to be absent; retrying...", exp.dnsName, exp.gatewayName)
+				expectationsMet = false
+				return false, nil
+			}
+		}
+		t.Logf("Continuing polling to ensure non-expected DNSRecords do not exist...")
```

Comment:
```suggestion
		if haveExpectNotPresent {
			t.Logf("Continuing polling to ensure non-expected DNSRecords do not exist...")
		}
```

To make it not log this misleadingly a single time when there is no negative expectations.